ygen/directory.go

```diff
@@ -135,9 +136,23 @@ func GetOrderedDirectories(directory map[string]*Directory) ([]string, map[strin
 // of a Directory. The Field is specified as a name in order to guarantee its
 // existence before processing.
 func FindSchemaPath(parent *Directory, fieldName string, absolutePaths bool) ([]string, error) {
+	return findSchemaPath(parent, fieldName, false, absolutePaths)
+}
+
+// findSchemaPath finds the relative or absolute schema path of a given field
+// of a Directory, or the shadowed field path (field duplicated and
```

Comment:
Do we need `shadowSchemaPaths` as an argument here. The `Directory` has been built with some particular behaviour in mind, so I wonder whether the right implementation here is:

- check whether this field was included in the `Directory`
- if it wasn't, then look for it in the shadowed fields.
- do the schema path resolution

I guess this is a bit harder to implement, since we need the paths for each `Entry` compared to this approach - but it might be neater as a way to be able to make `FindSchemaPath` have a simpler behaviour and not need to be given anything about the way that the IR was created in the future.

Reply:
I'm not quite seeing how this flag can be removed, because one of the new purposes of `FindSchemaPath` is to populate the new `shadow-path` struct tag, and without indicating to the function whether it wants the shadow or non-shadow path, it wouldn't know to do this.

For context see snippet from next PR (#429):
https://github.com/openconfig/ygot/blob/56a5401e16957ffda2481fcab16d15dbb91ef937/ygen/gogen.go#L1538-L1562

Reply:
Sorry for the delay replying here -- I'm a bit confused. The snippet you linked to doesn't seem to call `FindSchemaPath` right?

My first confusion is that this function is meant to be able to return either the shadow paths for a field that is within the view of the schema that we have - or its shadowed versions. For example, for `/foo/bar/config/baz` with `PreferIntendedConfig` on, then this should return `/foo/bar/config/baz` when `shadowSchemaPaths` is set to `true` and `/foo/bar/state/baz` when `shadowSchemaPaths` is set to `false`, right? In this model, we wouldn't have a way to be able to determine that within a particular directory an operational-state only field that has been removed is valid, I think -- which seems fine. This was my initial understanding of what we were trying to do (say whether any path was valid within the schema despite compression).

As for the exact implementation of the logic discussed above. AIUI, we're adding a new shadow path to the struct tags, which we might need to populate if this field happened to have both an intended and applied path. My assumption was that we would generate directories that either had `Directory{}.ShadowedFields` populated or not, and we wouldn't necessarily need to have this logic be in a common `findSchemaPath` implementation -- rather we would do something like:

* `findRealSchemaPaths` -> form `path` struct tag.
* `findShadowedSchemaPath` -> form `shadowPath` struct tag if the list is not-nil.

If the `ShadowedFields` map is empty then we would always get `nil` back from the 2nd function. I guess you're intending to have my two "real" and  "shadow" paths be one function with a flag? I think it'd be clearer as two - but I'm happy with improving the commenting such that we can say that `shadowSchemaPaths` actually changes the source of truth for this this function from the 'real' fields to those that were removed.

Reply:
Sorry for being not as clear as I could be. Please feel free to just have me explain again.

The snippet I linked calls `findMapPaths`, which in turn calls `findSchemaPath`:
https://github.com/openconfig/ygot/blob/56a5401e16957ffda2481fcab16d15dbb91ef937/ygen/gogen.go#L2083-L2088

`findMapPaths` is called twice during Go generation, once for retrieving the unshadowed leaf path, and once for retrieving the shadowed leaf path, which are then used to populate the struct tags. The boolean flag simply tells `findMapPaths` and `findSchemaPath` which tag is being populated so it can retrieve the correct path; therefore, if the shadow path didn't exist (say it's a state leaf), the function *should* return nothing. You can see how this is done in this test here (file from this PR): https://github.com/openconfig/ygot/blob/428adc619285f99ed99bf43561318be3ef37f249/ygen/genstate_test.go#L104-L177

Reply:
Makes sense - apologies for the delay in me understanding this!